barretenberg/cpp/src/barretenberg/vm2/simulation/bitwise.hpp

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <cstdint>
+
+#include "barretenberg/vm2/common/memory_types.hpp"
+#include "barretenberg/vm2/simulation/context.hpp"
+#include "barretenberg/vm2/simulation/events/bitwise_event.hpp"
+#include "barretenberg/vm2/simulation/events/event_emitter.hpp"
+#include "barretenberg/vm2/simulation/memory.hpp"
+
+namespace bb::avm2::simulation {
+
+class BitwiseInterface {
+  public:
+    virtual ~BitwiseInterface() = default;
+    virtual void op_and(ContextInterface&, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) = 0;
+    virtual void op_or(ContextInterface&, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) = 0;
+    virtual void op_xor(ContextInterface&, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) = 0;
+};
+
+class Bitwise : public BitwiseInterface {
+  public:
+    Bitwise(EventEmitterInterface<BitwiseEvent>& event_emitter)
+        : events(event_emitter)
+    {}
+
+    // Operands are expected to be direct.
+    void op_and(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) override;
+    void op_or(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) override;
+    void op_xor(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr) override;
+
+  private:
+    EventEmitterInterface<BitwiseEvent>& events;
```

Comment:
Please add a TODO to use "deduplicating events". We will not want to prove the ops many times. I will set up infra for this so don't do it now.